juju/juju
# Fix undertaker restarting when parameters changed.

- This fixes the restarting of the undertaker worker when the destroy-model parameters have change.
- Adds a missing test to ensure this behaviour.
- Updates juju/worker to v3.4.0 for catacomb context.

## QA steps

1. Bootstrap k8s.
2. Add a model.
3. Deploy a simple shell charm that sleeps in the remove hook for a while.
4. Destroy model without force.
```
controller-0: 12:02:04 INFO juju.worker.undertaker destroying model "a"
```
5. Destroy model with `--force --timeout 10s`
```
controller-0: 12:02:58 INFO juju.worker.undertaker model destroy parameters changed: restarting undertaker worker
controller-0: 12:02:58 ERROR juju.worker.undertaker destroy model failed: process dying model: context canceled
controller-0: 12:03:02 INFO juju.worker.undertaker force destroying model "a" with timeout 10s
```
6. Model should be deleted after a short while.

```
$ cd worker/undertaker
$ go test -c -race
$ go run golang.org/x/tools/cmd/stress ./undertaker.test -check.f=TestExitOnModelChanged
5s: 776 runs so far, 0 failures
10s: 1620 runs so far, 0 failures
```

## Documentation changes

N/A

## Bug reference

https://bugs.launchpad.net/juju/+bug/2034643
